pkg/metainfo/kvmetainfo/stream.go

```diff
@@ -62,17 +63,40 @@ func (stream *readonlyStream) segment(ctx context.Context, index int64) (info st
 		copy(segment.EncryptedKeyNonce[:], segmentMeta.KeyNonce)
 		segment.EncryptedKey = segmentMeta.EncryptedKey
 	} else {
+		segmentPath = storj.JoinPaths("l", stream.encryptedPath)
 		segment.Size = stream.info.LastSegment.Size
 		segment.EncryptedKeyNonce = stream.info.LastSegment.EncryptedKeyNonce
 		segment.EncryptedKey = stream.info.LastSegment.EncryptedKey
 	}
 
-	var nonce storj.Nonce
-	_, err = encryption.Increment(&nonce, index+1)
+	contentKey, err := encryption.DecryptKey(segment.EncryptedKey, stream.Info().EncryptionScheme.Cipher, stream.streamKey, segment.EncryptedKeyNonce)
 	if err != nil {
 		return segment, err
 	}
 
+	nonce := new(storj.Nonce)
+	_, err = encryption.Increment(nonce, index+1)
+	if err != nil {
+		return segment, err
+	}
+
+	pointer, _, err := stream.db.pointers.Get(ctx, segmentPath)
+	if err != nil {
+		return segment, err
+	}
+
+	if pointer.GetType() == pb.Pointer_INLINE {
+		segment.Inline, err = encryption.Decrypt(pointer.InlineSegment, stream.Info().EncryptionScheme.Cipher, contentKey, nonce)
+	} else {
+		segment.PieceID = storj.PieceID(pointer.Remote.PieceId)
+		segment.Pieces = make([]storj.Piece, 0, len(pointer.Remote.RemotePieces))
+		for _, piece := range pointer.Remote.RemotePieces {
+			var nodeID storj.NodeID
+			copy(nodeID[:], piece.NodeId)
```

Comment:
@egonelbre Can we have a better way to initialize `storj.NodeID`?

Reply:
At the moment no. Also currently storj.NodeID is not yet compatible with yet piece location, since the location uses more bytes.